uk.co.compucorp.civicrm.hrleaveandabsences/tests/phpunit/CRM/HRLeaveAndAbsences/BAO/LeaveRequestTest.php

```diff
@@ -319,4 +322,326 @@ public function testGetBalanceChangeByAbsenceTypeCanReturnTheBalanceForOnlyPubli
+    PublicHolidayLeaveRequestFabricator::fabricate($contact['id'], $publicHoliday);
+
+    $contactId = $contact['id'];
+    $fromDate = date("Y-11-05");
+    $toDate = date("Y-11-08");
+    $fromType = '1/2 AM';
+    $toType = 'All Day';
+
+    $startDate = new DateTime($fromDate);
+    $endDate = new DateTime($toDate);
+    $endDateUnmodified = new DateTime($toDate);
+    // add one day to end date to include it in DatePeriod
+    $endDate->modify('+1 day');
+    $interval   = new DateInterval('P1D');
+    $datePeriod = new DatePeriod($startDate, $interval, $endDate);
+    $expectedResultsBreakdown = [
+      'amount' => 0,
+      'breakdown' => []
+    ];
+    $fromDateIsHalfDay = in_array($fromType, ['1/2 AM', '1/2 PM']);
+    $toDateIsHalfDay = in_array($toType, ['1/2 AM', '1/2 PM']);
+
+    //mimick how LeaveRequest::calculateBalanceChange will handle this to get an expected result
+    foreach ($datePeriod as $date) {
+      if(in_array($date->format('N'), [6, 7])){
+        $amount = 0;
+        $type = 'Weekend';
+      }
+      else{
+        $amount = -1 * 1.0;
+        $type = 'All day';
+      }
+
+      if($fromDateIsHalfDay && $date->format('Y-m-d') == $startDate->format('Y-m-d') && $amount != 0) {
+        $amount = -1 * 0.5;
+        $type =  $fromType;
+      }
+
+      if($toDateIsHalfDay && $date->format('Y-m-d') == $endDateUnmodified->format('Y-m-d') && $amount != 0) {
+        $amount = -1 * 0.5;
+        $type =  $fromType;
+      }
+
+      //For this day, LeaveRequest::calculateBalanceChange should see it as a public holiday
+      if($date->format('Y-m-d') == date('Y-11-07')) {
+        $amount = 0;
+        $type =  'Public Holiday';
+      }
+
+      $result = [
+        'date' => $date->format('Y-m-d'),
+        'amount' => abs($amount),
+        'type' => $type
+      ];
+      $expectedResultsBreakdown['amount'] += $amount;
+      $expectedResultsBreakdown['breakdown'][] = $result;
+
+    }
+    $result = LeaveRequest::calculateBalanceChange($contactId, $fromDate, $fromType, $toDate, $toType);
+    $this->assertEquals($expectedResultsBreakdown, $result);
+  }
+
+  public function testCalculateBalanceChangeForALeaveRequestForAContactWithMultipleWeeks() {
+    $contact = ContactFabricator::fabricate();
+    $periodStartDate = date('Y-01-01');
+    $title = 'Job Title';
+
+    HRJobContractFabricator::fabricate([
+      'contact_id' => $contact['id']
+    ],
+      [
+        'period_start_date' => $periodStartDate,
+        'title' => $title
+      ]);
+
+    // Week 1 weekdays: monday, wednesday and friday
+    // Week 2 weekdays: tuesday and thursday
+    $pattern = WorkPatternFabricator::fabricateWithTwoWeeksAnd31AndHalfHours();
+    //attach the work pattern to the contact
+    ContactWorkPatternFabricator::fabricate([
+      'contact_id' => $contact['id'],
+      'pattern_id' => $pattern->id
+    ]);
+
+    $contactId = $contact['id'];
```

Comment:
do you really need this? couldn't you just use $contact['id'] to pass that id to the other methods?